### PR TITLE
Add MSFT as auth provider

### PR DESCRIPTION
**Pull request by bencooper222**

### CODE REVIEW

File: src/server/auth/microsoft.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+export const strategy = (models: Models): Strategy =>
+	new Strategy(
+		{
+			allowHttpForRedirectUrl: true,
+			clientID: MSFT_CLIENT_ID,
+			clientSecret: MSFT_CLIENT_SECRET,
+			identityMetadata: MSFT_TENANT_NAME,
+			loggingLevel: 'info',
+			passReqToCallback: false,
+			redirectUrl: `http://localhost:8081${MSFT_REDIRECT_URL}`, // TODO: FIX
```
Comment (bencooper222) [inline src/server/auth/microsoft.ts]: Is there a good way for me to have the env var be `MSFT_REDIRECT_URL="/api/auth/microsoft/callback"` and still make the path relative to the deploy context?
`

**Comment by bencooper222** [comment]
Bleh given's microsoft's bullshit, I should probably make a separate verifycallback function I think

File: src/server/auth/microsoft.ts (hunk 1 above)
Comment (leonm1) [reply src/server/auth/microsoft.ts]: GitHub doesn't mind a relative redirect path, but Google doesn't allow it, so mostly I've just been setting the full path in the env var

File: src/server/auth/helpers.ts [hunk 2]
```diff
@@ -18,9 +18,11 @@ export const verifyCallback = async (
 	})) || { userType: null };
 
 	try {
-		const { emails: [{ value: email }] = [{ value: null }] } = profile;
+		let { emails: [{ value: email }] = [{ value: null }] } = profile;
 		if (email == null) {
-			throw new Error(`Email not provided by provider ${profile}`);
+			// microsoft is fucking stupid, I can't even.
+			email = profile._json.email; // eslint-disable-line
+			if (email == null) throw new Error(`Email not provided by provider ${profile}`);
```
Comment (leonm1) [inline src/server/auth/helpers.ts]: LMAO

But please remove profanity from code. Also should rebase because new logging code added.

**Comment by leonm1** [comment]
> Bleh given's microsoft's bullshit, I should probably make a separate verifyCallback function I think - thoughts appreciated

The only thing I'm considering is guarding the assignment statement inside an if condition in case one of the other providers comes into a similar situation

**Comment by cktang88** [comment]
^^ as long as CI passes

File: src/server/auth/microsoft.ts (hunk 1 above)
Comment (bencooper222) [reply src/server/auth/microsoft.ts]: Gotcha. Also, how aggressive should the loggingLevel be - `info`, `warn` or `error` (or none). And should I log PII?

Comment (cktang88) [reply src/server/auth/microsoft.ts]: info if no err, error if err, warn has no use case

**Comment by bencooper222** [comment]
I think I'm just going to create a separate verifyMicrosoftCallback. I don't think there's a good way to make typecheck pass without doing that because I have to get both email and id (or, as microsoft calls it "sub" because of course they do) in a special way and the if statement gating to get that to work (along with a different Profile type) just isn't worth it imo

**Comment by bencooper222** [comment]
Short of exhaustively testing all the oauth providers, I think this is fine. One last review @leonm1 - and please note the env var requirements before moving to prod

**Comment by cktang88** [comment]
squash merge?

**Pull request merged**